### PR TITLE
Validator/authority contract

### PR DESCRIPTION
**Pull request by keorn**
Ability to use a contract at a specified address to determine the current list of validators (renamed from authorities) as an alternative to fixed spec list. The test contract is [here](https://github.com/ethcore/contracts/pull/27).

<s>Another change is that `Engine` now holds a `Weak<Client>` rather than `IoChannel`.</s>

<s>Do not merge for 1.5, goes after #3932.<\s>

### CODE REVIEW

**Comment by gavofyork** [comment]
> `Engine` now holds a `Weak<Client>` rather than `IoChannel`

does this not compromise IPC? @NikVolf ?

File: ethcore/src/engines/validator_set/contract.rs [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+
+use std::sync::Weak;
+use util::*;
+use client::{Client, BlockChainClient};
+use client::chain_notify::ChainNotify;
+use super::ValidatorSet;
+use super::simple_list::SimpleList;
+
+/// The validator contract should have the following interface:
+/// [{"constant":true,"inputs":[],"name":"get_validators","outputs":[{"name":"","type":"address[]"}],"payable":false,"type":"function"}]
```
Comment (gavofyork) [inline ethcore/src/engines/validator_set/contract.rs]: Solidity functions are camelCase by convention (including in all our Soldity contracts).

**Comment by rphmeier** [comment]
> Another change is that Engine now holds a Weak<Client> rather than IoChannel.

I really don't agree with this -- Engine is supposed to be a relatively agnostic consensus mechanism which the client and other related services can leverage. Any communication with other parts of the codebase should be done through a more general mechanism. I'm not sure `IoChannel` is quite right, but requiring a dependency on `Client` in the engine is overly restrictive.

**Comment by gavofyork** [comment]
agree, `Engine` is meant to be an resource, not a coordinator.

**Comment by keorn** [comment]
Now the two `Engine`s hold `EngineClient`.
`Client` is passed since when necessary (validator set is in the state) an `Engine` will `add_notify` which inspects the validators using a closure, similar to the updater.

File: ethcore/src/miner/miner.rs [hunk 2]
```diff
@@ -764,9 +764,11 @@ impl MinerService for Miner {
 			if let Some(ref ap) = self.accounts {
 				ap.sign(address.clone(), Some(password.clone()), Default::default())?;
 			}
-			let mut sealing_work = self.sealing_work.lock();
-			sealing_work.enabled = self.engine.is_sealer(&address).unwrap_or(false);
-			*self.author.write() = address;
+			{
```
Comment (gavofyork) [inline ethcore/src/miner/miner.rs]: good to have a comment here to ensure any newcomer understands what the parens are for.

**Comment by gavofyork** [comment]
minor grumble. ltgm otherwise.

**Pull request merged**